### PR TITLE
Code Review: Update Johnnie’s text utilities plugin to use the api (#10574)

### PR DESCRIPTION
**Pull request by samdeane**
Code review for Update Johnnie’s text utilities plugin to use the api (#10574):

> I'm finding that the best way to drive the development of the JS API is to try to make things using it.
> 
> @randomsequence has a text utilities plugin which is a bit broken - seems like it might be a good candidate for rewriting to use the API, to see what emerges from that.

Connect to BohemianCoding/Sketch#10574.

### CODE REVIEW

File: Source/Layer.js [hunk 1]
```diff
@@ -7,317 +7,405 @@
...
-    */
+  All Layer objects respond to this method, but only shape layers (rectangles, ovals, paths etc) return true.
 
-    get isShape() { return false; }
+  @return {bool} true for instances of Group, false for any other layer type.
+  */
 
-    /**
-        Is this an image layer?
+  get isShape() { return false; }
```
Comment (pieteromvlee) [inline Source/Layer.js]: Is this not something you'd put on a base class and then query our traits?

File: Source/Layer.js [hunk 2]
```diff
@@ -7,317 +7,405 @@
...
+  /**
+  Select the layer.
+  This will clear the previous selection. Use addToSelection() if you wish
+  to preserve the existing selection.
+  */
 
-        @param {dictionary} properties Properties to apply to the shape.
-        @return {Shape} the new shape.
-    */
+  select() {
```
Comment (pieteromvlee) [inline Source/Layer.js]: `select` vs `deselect` behaviour fels inconsistent with how they treat deselecting others layers?

Comment (pieteromvlee) [inline Source/Layer.js]: Or maybe its fine. Probably is... Ignore me

File: Source/Layer.js (hunk 1 above)
Comment (samdeane) [inline Source/Layer.js]: Could do, but this applies to the javascript classes themselves, not to the underlying traits. I'm deliberately simplifying the view in the API, so I didn't necessarily want to expose the traits.

Comment (samdeane) [inline Source/Layer.js]: I can always change the implementation later.

File: Source/Group.js [hunk 3]
```diff
@@ -6,49 +6,112 @@
...
+
+  /**
+  Is this an group?
+
+  All Layer objects respond to this method, but only Groups or things that inherit from groups return true.
+
+  @return {bool} true for instances of Group, false for any other layer type.
+  */
+
+  is_group() {
```
Comment (bomberstudios) [inline Source/Group.js]: Why is this `is_group` when you're using `isGroup` a few lines further down?

File: Source/Layer.js [hunk 4]
```diff
@@ -7,317 +7,405 @@
...
-    */
+  All Layer objects respond to this method, but only Groups or things that inherit from groups return true.
 
-    get isGroup() { return false; }
+  @return {bool} true for instances of Group, false for any other layer type.
+  */
 
-    /**
-        Is this a text layer?
+  get isGroup() { return false; }
```
Comment (bomberstudios) [inline Source/Layer.js]: See comment above

File: Source/Layer.js [hunk 5]
```diff
@@ -7,317 +7,405 @@
...
+    if (!style) {
+      style = new Style()
     }
 
-    /**
-      Extract the frame to use for a layer from some properties.
-      If the frame wasn't supplied in the properties, we return a default value instead.
+    var fills = properties.fills
+    if (fills) {
+      delete properties["fills"]
```
Comment (bomberstudios) [inline Source/Layer.js]: I'm confused by this `properties.fills` vs `properties["fills"]`, but something tells me this is one of those stupid things JavaScript does, so feel free to ignore the comment

File: Source/Style.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+// ********************************
+
+import { WrappedObject } from './WrappedObject.js'
+
+/// A solid fill/border.
+const BCFillTypeColor    = 0
+
+/// A gradient fill/border.
+const BCFillTypeGradient = 1
+
```
Comment (bomberstudios) [inline Source/Style.js]: I guess 2 and 3 are Radial and Angular gradients? Are they missing for any particular reason?

File: Source/Style.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+      and other more complex border parameters to be specified.
+
+      @param {array} values A list of colors - each one representing a border to create.
+
+    */
+
+    set borders(value) {
+      var objects = []
+      for (var b in value) {
+        var color = MSColor.colorWithSVGString_(value[b])
```
Comment (bomberstudios) [inline Source/Style.js]: Leaving this here to plant a seed in your mind: we need a nice JS API to work with colours : )

File: Source/Group.js (hunk 3 above)
Comment (samdeane) [inline Source/Group.js]: thanks, that's a mistake. I think it's been fixed in the unit test branch!

File: Source/Style.js (hunk 6 above)
Comment (samdeane) [inline Source/Style.js]: They are missing right now because the API hasn't implemented them. I'll add them later.

File: Source/Style.js (hunk 7 above)
Comment (samdeane) [inline Source/Style.js]: 👍

**Comment by samdeane** [comment]
> Maybe traits are not accessible in the JS bridge, but those methods can be written on the base class I think so that you dont have to repeat them for each subclass

`Layer` *is* the base class in this case.

**Pull request merged**